packages/chain-adapters/src/cosmossdk/CosmosSdkBaseAdapter.ts

```diff
@@ -0,0 +1,221 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+// no-unused-vars is temporarily disabled until more functions are implemented
+import { CAIP2, caip2, caip19 } from '@shapeshiftoss/caip'
+import { AssetNamespace } from '@shapeshiftoss/caip/dist/caip19/caip19'
+import { CosmosSignTx } from '@shapeshiftoss/hdwallet-core'
+import { BIP44Params, chainAdapters, ChainTypes } from '@shapeshiftoss/types'
+import { cosmos } from '@shapeshiftoss/unchained-client'
+import WAValidator from 'multicoin-address-validator'
+
+import { ChainAdapter as IChainAdapter } from '../api'
+import { ChainAdapter } from '../bitcoin'
+import { ErrorHandler } from '../error/ErrorHandler'
+
+export type CosmosChainTypes = ChainTypes.Cosmos | ChainTypes.Osmosis
+
+export interface ChainAdapterArgs {
+  chainId?: CAIP2
+  providers: {
+    http: cosmos.api.V1Api
+    // unchained-client 5.1.1 does not have a websocket client for cosmos
+  }
+  symbol: string
+}
+
+export abstract class CosmosSdkBaseAdapter<T extends CosmosChainTypes> implements IChainAdapter<T> {
+  protected readonly chainId: CAIP2
+  protected readonly supportedChainIds: CAIP2[]
+  protected readonly symbol: string
+  protected readonly providers: {
+    http: cosmos.api.V1Api
+  }
+
+  public static readonly defaultBIP44Params: BIP44Params = {
+    purpose: 44,
+    coinType: 118,
+    accountNumber: 0
+  }
+
+  protected constructor(args: ChainAdapterArgs) {
+    if (args.chainId && this.supportedChainIds.includes(args.chainId)) {
+      this.chainId = args.chainId
+    }
+    this.providers = args.providers
+  }
+
+  getType(): T {
+    throw new Error('Method not implemented.')
+  }
+
+  getChainId(): CAIP2 {
+    return this.chainId
+  }
+
+  getCaip2(): CAIP2 {
+    return this.chainId
+  }
+
+  async getAccount(pubkey: string): Promise<chainAdapters.Account<T>> {
+    try {
+      const caip = await this.getCaip2()
+      const { chain, network } = caip2.fromCAIP2(caip)
+      const { data } = await this.providers.http.getAccount({ pubkey })
+
+      return {
+        balance: data.balance,
+        caip2: caip,
+        // This is the caip19 for default asset on the chain (ATOM/OSMO/etc)
+        caip19: caip19.toCAIP19({
+          chain,
+          network,
+          assetNamespace: AssetNamespace.Slip44,
+          assetReference: '118'
+        }),
+        chain: this.getType(),
+        chainSpecific: {
+          sequence: data.sequence
+        },
+        pubkey: data.pubkey
+      } as chainAdapters.Account<T>
```

Comment:
Seems like we have some issues with returned data and/or generic type definitions

Reply:
we added a comment to try to explain it. It's the fact that T is generic and the compiler can't guarantee that the `chainSpecific` object defined here is correctly defined. If we used a specific `ChainTypes` enum value to define the return type, then it would be fine.

I spent a while trying to figure out some way around this but there isn't based on how we currently define `ChainSpecifc`. I don't think it's worth fighting with right now since we can manually verify that this is correct, especially if we have unit tests to verify it.